### DISCUSSION TITLE
Target-Typed New Idea

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#1709] [Language Ideas]
Hi All,
As I was reading:
https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-06-25.md

I had an idea regarding constructors and such.  I generally really like using the initializer syntax because it gives me a clear view of the property->value mapping.  I've wished this syntax was extended to regular constructors.

For example, given the following class definition:
```
public class Person {
  public string FirstName {get; private set;}
  public string LastName {get; private set;}

  public int Age {get; set; }
  public Person(string FirstName, string LastName){
    this.FirstName = FirstName;
    this.LastName = LastName;
  }

 public Person(string Name){
   var Names = string.Split(Name, " ");
   this.FirstName = Names.FirstOrDefault();
   this.LastName = Names.LastOrDefault();
 }

}
````

I want to be able to new up a person as follows:
````
var P1 = new Person {
  Name = "Bob Smith",
  Age = 21,
};
//This would be the same as:
var P1 = new Person("Bob Smith) {
  Age = 21,
};



var P2 = new Person {
  FirstName = "Bob",
  LastName = "Smith",
  Age = 21,
};
//This would be the same as:
var P2 = new Person("Bob", "Smith") {
  Age = 21,
};
````
Inside of an initializer, I should be able to specify constructor parameters as if they were properties.
If I'm specifying constructor parameters, I must specify all the parameters for an overload ortherwise it is an error.  I'm thinking something similar to the way that PowerShell does parameter sets would be really useful.

**Comment by HaloFour** [comment]
Would this only apply when a constructor argument name matches a property name?  Would that property have to be read-only?

**Comment by TonyValenti** [comment]
The constructor arguments should not be required to match a property name.  For example, I should be able to do the following:
````
public class Person {

 public string FirstName {get; set;}
 public string LastName {get; set;}
 public int Age {get; set;}

  public Person(string Name_First, string Name_Last){
   this.FirstName = Name_First;
    this.LastName = Name_Last;
  }


}
...

var P = new Person {
  Name_First = "Bob",
  Name_Last = "Smith",  //These two are required and get passed to the constructor.
  Age = 21,
  FirstName = "Bob" //This is redundant but it calls the property accessor.
};

````

Also, in the example above, the Name_First and Name_Last values would have to be specified since there is no default constructor.
